src/Widget.php

```diff
@@ -0,0 +1,182 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Widget;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Yiisoft\Widget\Exception\InvalidConfigException;
+use Yiisoft\Widget\Event\AfterRun;
+use Yiisoft\Widget\Event\BeforeRun;
+use Yiisoft\Widget\Factory\WidgetFactory;
+
+/**
+ * Widget is the base class for widgets.
+ */
+class Widget
+{
+    /**
+     * @var EventDispatcherInterface $eventDispatcher
+     */
+    protected EventDispatcherInterface $eventDispatcher;
+
+    /**
+     * The widgets that are currently being rendered (not ended). This property is maintained by {@see begin()} and
+     * {@see end} methods.
+     *
+     * @var array $stack
+     */
+    protected static array $stack;
+
+    /**
+     * @var Widget $widget
+     */
+    protected static Widget $widget;
+
+    public function __construct(EventDispatcherInterface $eventDispatcher)
+    {
+        $this->eventDispatcher = $eventDispatcher;
+    }
+
+    public function init(): void
+    {
+    }
+
+    public function run(): string
```

Comment:
Should it be called directly? I guess not because "before run" and "after run" events won't be called then. Let's make it `protected`.

Reply:
Done.